### PR TITLE
Save String allocations by freezing ❄️ them (3.0 series)

### PR DESCRIPTION
**Pull request by elia**
Inspired by the great work done by @schneems on rails.

Here are some numbers taken from the current (Spree) app I’m working on:

```
'file-digest'   39k > 11k
'\1'            11k > 59
'NilClass'      10k > (less than 24)
```

PS. if this is ok I'll port the changes to master too

### CODE REVIEW

**Comment by rafaelfranca** [comment]
It lacks some numbers about how this improved the overall speed of the application. I can see that it saves a lot of objects, but it still don't justify this change.

**Comment by elia** [comment]
I'll try to demonstrate the value through benchmarks, but that could take a while (on my part 😕).

Of course this change is not going to make an app 2x faster and would probably account only for a tiny improvement. Nonetheless I think for a library (vs. application) the numbers shown should be a reason good enough. I found an example of such changes in commit a6bf1f97236283e89769f6af59f3b48a297b738b by @tenderlove.

Would it be better if I reduce the scope of the change to just the three cases I shown the numbers for?

**Comment by rafaelfranca** [comment]
> Would it be better if I reduce the scope of the change to just the three cases I shown the numbers for?

Yes. We don't want to introduce `.freeze` calls in the entire codebase just because it is "faster".

**Comment by tenderlove** [comment]
TBH, I think we can actually eliminate that method in digest utils if we use objects and polymorphism.  IIRC, it generates a cache key that should not change, so I'm not sure that adding freezes will help.  But OTOH, I could be totally wrong and we're calling that method a bunch.  If I *am* wrong, then that is even more motivation to move to polymorphism and eliminate that method.

**Comment by elia** [comment]
With some [encouragement](https://github.com/elia/sprockets/blob/3-0/let-it-go/lib/sprockets/digest_utils.rb#L37-L44) on doing microbenchmarks here's the results for the plan vs. frozen implementations of the `#digest` method ([code here](https://gist.github.com/elia/e63d3f3dcf1ac0850483)):

```
Calculating -------------------------------------
              frozen     5.682k i/100ms
               plain     5.249k i/100ms
          nil-frozen     5.453k i/100ms
           nil-plain     4.743k i/100ms
-------------------------------------------------
              frozen     51.649k (± 3.5%) i/s -    261.372k
               plain     45.329k (±11.9%) i/s -    225.707k
          nil-frozen     44.550k (± 8.8%) i/s -    223.573k
           nil-plain     37.685k (± 3.4%) i/s -    189.720k

Comparison:
              frozen:    51649.0 i/s
               plain:    45329.4 i/s - 1.14x slower
          nil-frozen:    44549.6 i/s - 1.16x slower
           nil-plain:    37684.6 i/s - 1.37x slower
```

The `nil-` ones are passed an array of `nil`s.


---

Looking again at the source thos I wonder if the `#hash` value of the root object can be used instead along maybe with the name of the root object's class …or the `#hash` value of the individual objects. Just throwing ideas…

**Comment by elia** [comment]
Nay, frozen is still the fastest:

```
Calculating -------------------------------------
              frozen     5.740k i/100ms
               plain     5.377k i/100ms
                hash     4.568k i/100ms
          nil-frozen     5.376k i/100ms
           nil-plain     4.933k i/100ms
            nil-hash     4.257k i/100ms
-------------------------------------------------
              frozen     50.111k (± 4.7%) i/s -    252.560k
               plain     44.650k (± 4.3%) i/s -    225.834k
                hash     38.526k (± 3.1%) i/s -    196.424k
          nil-frozen     45.129k (± 3.2%) i/s -    231.168k
           nil-plain     36.449k (±10.6%) i/s -    182.521k
            nil-hash     31.716k (± 4.4%) i/s -    161.766k

Comparison:
              frozen:    50110.8 i/s
          nil-frozen:    45129.5 i/s - 1.11x slower
               plain:    44650.0 i/s - 1.12x slower
                hash:    38525.7 i/s - 1.30x slower
           nil-plain:    36448.7 i/s - 1.37x slower
            nil-hash:    31715.7 i/s - 1.58x slower
```

The implementation of the hash based digest I used it this:

```ruby
def hash_digest(obj)
  digest = digest_class.new
  digest << obj.class.name
  digest << obj.hash.to_s
  digest.digest
end
```

**Comment by elia** [comment]
@rafaelfranca I've updated the PR removing the `.freeze` calls not covered by data.

File: lib/sprockets/dependencies.rb [hunk 1]
```diff
@@ -62,7 +62,17 @@ def resolve_dependencies(uris)
     #
     # Returns resolved Object.
     def resolve_dependency(str)
-      scheme = str[/([^:]+)/, 1]
+      file_digest        = 'file-digest'.freeze
```
Comment (rafaelfranca) [inline lib/sprockets/dependencies.rb]: Can we inline there variables?

**Comment by rafaelfranca** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

File: lib/sprockets/dependencies.rb (hunk 1 above)
Comment (elia) [inline lib/sprockets/dependencies.rb]: Good call :)

**Comment by rafaelfranca** [comment]
> that is even more motivation to move to polymorphism and eliminate that method.

:+1:

**Pull request merged**